pupil_src/shared_modules/video_capture/uvc_backend.py

```diff
@@ -773,42 +774,58 @@ def dev_selection_list():
             ]
             return zip(*dev_pairs)
 
-        def activate(source_uid):
-            if not source_uid:
-                return
-            if not uvc.is_accessible(source_uid):
-                logger.error("The selected camera is already in use or blocked.")
-                return
-            settings = {
-                "frame_size": self.g_pool.capture.frame_size,
-                "frame_rate": self.g_pool.capture.frame_rate,
-                "uid": source_uid,
-            }
-            if self.g_pool.process == "world":
-                self.notify_all(
-                    {"subject": "start_plugin", "name": "UVC_Source", "args": settings}
-                )
-            else:
-                self.notify_all(
-                    {
-                        "subject": "start_eye_capture",
-                        "target": self.g_pool.process,
-                        "name": "UVC_Source",
-                        "args": settings,
-                    }
-                )
-
         ui_elements.append(
             ui.Selector(
                 "selected_source",
                 selection_getter=dev_selection_list,
                 getter=lambda: None,
-                setter=activate,
+                setter=self.activate,
                 label="Activate source",
             )
         )
         self.menu.extend(ui_elements)
 
+    def activate(self, source_uid):
+        if not source_uid:
+            return
+        if not uvc.is_accessible(source_uid):
+            logger.error("The selected camera is already in use or blocked.")
+            return
+        settings = {
+            "frame_size": self.g_pool.capture.frame_size,
+            "frame_rate": self.g_pool.capture.frame_rate,
+            "uid": source_uid,
+        }
+        if self.g_pool.process == "world":
+            self.notify_all(
+                {"subject": "start_plugin", "name": "UVC_Source", "args": settings}
+            )
+        else:
+            self.notify_all(
+                {
+                    "subject": "start_eye_capture",
+                    "target": self.g_pool.process,
+                    "name": "UVC_Source",
+                    "args": settings,
+                }
+            )
+
+    def auto_activate_source(self):
+        if not self.devices or len(self.devices) == 0:
+            logger.warning("No UVC source available.")
+            return
+
+        cam_selection_lut = {"eye0": "ID0", "eye1": "ID1", "world": "ID2"}
+        cam_id = cam_selection_lut[self.g_pool.process]
+        source_id = [d["uid"] for d in self.devices if cam_id in d["name"]]
+        if len(source_id) != 1:
+            logger.warning("Unexpected process name / camera ID pair")
```

Comment:
Maybe include the `cam_id` you are looking for in the warning?

Reply:
Better yet, I incorporated what you suggested for the `ndsi_backend` for both classes:

```python
        try:
            source_id = next(
                uid for uid, name in self.devices.items() if cam_id in name
            )
        except StopIteration:
            logger.warning("No camera found with ID: {}".format(cam_id))
            source_id = None
        self.activate(source_id)
```